go-controller/pkg/ovn/gateway.go

```diff
@@ -760,3 +763,167 @@ func (gw *GatewayManager) addExternalSwitch(prefix, interfaceID, nodeName, gatew
 
 	return nil
 }
+
+// Cleanup removes all the NB DB objects created for a node's gateway
+func (gw *GatewayManager) Cleanup() error {
+	// Get the gateway router port's IP address (connected to join switch)
+	var nextHops []net.IP
+
+	gwIPAddrs, err := libovsdbutil.GetLRPAddrs(gw.nbClient, types.GWRouterToJoinSwitchPrefix+gw.gwRouterName)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return err
+	}
+
+	for _, gwIPAddr := range gwIPAddrs {
+		nextHops = append(nextHops, gwIPAddr.IP)
+	}
+	gw.staticRouteCleanup(nextHops)
+	gw.policyRouteCleanup(nextHops)
+
+	// Remove the patch port that connects join switch to gateway router
+	portName := types.JoinSwitchToGWRouterPrefix + gw.gwRouterName
+	lsp := nbdb.LogicalSwitchPort{Name: portName}
+	sw := nbdb.LogicalSwitch{Name: gw.transitSwitchName}
+	err = libovsdbops.DeleteLogicalSwitchPorts(gw.nbClient, &sw, &lsp)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete logical switch port %s from switch %s: %w", portName, sw.Name, err)
+	}
+
+	// Remove the logical router port on the gateway router that connects to the join switch
+	logicalRouter := nbdb.LogicalRouter{Name: gw.gwRouterName}
+	logicalRouterPort := nbdb.LogicalRouterPort{
+		Name: types.GWRouterToJoinSwitchPrefix + gw.gwRouterName,
+	}
+	err = libovsdbops.DeleteLogicalRouterPorts(gw.nbClient, &logicalRouter, &logicalRouterPort)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete port %s on router %s: %w", logicalRouterPort.Name, gw.gwRouterName, err)
+	}
+
+	// Remove the static mac bindings of the gateway router
+	err = gateway.DeleteDummyGWMacBindings(gw.nbClient, gw.gwRouterName)
+	if err != nil {
+		return fmt.Errorf("failed to delete GR dummy mac bindings for node %s: %w", gw.nodeName, err)
+	}
+
+	// Remove the gateway router associated with nodeName
+	err = libovsdbops.DeleteLogicalRouter(gw.nbClient, &logicalRouter)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete gateway router %s: %w", gw.gwRouterName, err)
+	}
+
+	// Remove external switch
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, gw.extSwitchName)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", gw.extSwitchName, err)
+	}
+
+	exGWexternalSwitch := types.EgressGWSwitchPrefix + gw.extSwitchName
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, exGWexternalSwitch)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", exGWexternalSwitch, err)
+	}
+
+	// This will cleanup the NodeSubnetPolicy in local and shared gateway modes. It will be a no-op for any other mode.
+	gw.delPbrAndNatRules(gw.nodeName, nil)
+	return nil
+}
+
+func (gw *GatewayManager) delPbrAndNatRules(nodeName string, lrpTypes []string) {
```

Comment:
remove second func arg - you always pass nil. Its confusing as well. Below for `removeLRPolicies` it accepts this arg but its called `priorities` there where as here its called `lrpTypes`